server/services/v1/auth/gotrue.go

```diff
@@ -0,0 +1,601 @@
+
+	currentSub, err := GetCurrentSub(ctx)
+	if err != nil {
+		return nil, errors.Internal("Failed to list applications: reason = %s", err.Error())
+	}
+
+	currentNamespace, err := request.GetNamespace(ctx)
+	if err != nil {
+		return nil, errors.Internal("Failed to list applications: reason = %s", err.Error())
+	}
+
+	// make gotrue call
+	creationTime := time.Now().UnixMilli()
+	err = createUser(ctx, clientId, clientSecret, currentSub, currentNamespace, req.GetName(), req.GetDescription(), req.GetProject(), creationTime, g)
+	if err != nil {
+		return nil, err
+	}
+
+	return &api.CreateAppKeyResponse{
+		CreatedAppKey: &api.AppKey{
+			Id:          clientId,
+			Name:        req.GetName(),
+			Description: req.GetDescription(),
+			Secret:      clientSecret,
+			CreatedAt:   creationTime,
+			CreatedBy:   currentSub,
+			Project:     req.GetProject(),
+		},
+	}, nil
+}
+
+func (g *gotrue) UpdateAppKey(ctx context.Context, req *api.UpdateAppKeyRequest) (*api.UpdateAppKeyResponse, error) {
+	email := fmt.Sprintf("%s%s", req.GetId(), g.AuthConfig.GotrueUsernameSuffix)
+	updateAppKeyUrl := fmt.Sprintf("%s/admin/users/%s", g.AuthConfig.GotrueURL, email)
+
+	currentSub, err := GetCurrentSub(ctx)
+	if err != nil {
+		log.Err(err).Msg("Couldn't resolve current sub")
+		return nil, errors.Internal("Failed to update app key")
+	}
+
+	newAppMetadata := UserAppData{
+		UpdatedAt: time.Now().UnixMilli(),
+		UpdatedBy: currentSub,
+	}
+
+	if req.GetName() != "" {
+		newAppMetadata.Name = req.GetName()
+	}
+
+	if req.GetDescription() != "" {
+		newAppMetadata.Description = req.GetDescription()
+	}
+	appMetadataMap := make(map[string]UserAppData)
+	appMetadataMap["app_metadata"] = newAppMetadata
+	payloadBytes, err := jsoniter.Marshal(appMetadataMap)
+	if err != nil {
+		log.Err(err).Msg("Failed to marshal payload")
+		return nil, errors.Internal("Unable to update app key")
+	}
+	payloadBytesReader := bytes.NewReader(payloadBytes)
+
+	client := &http.Client{}
+	updateAppKeyReq, err := http.NewRequestWithContext(ctx, http.MethodPut, updateAppKeyUrl, payloadBytesReader)
+	if err != nil {
+		log.Err(err).Msg("Failed to construct updateAppKeyReq")
+		return nil, errors.Internal("Unable to update app key")
+	}
+	adminAccessToken, _, err := getGotrueAdminAccessToken(ctx, g)
+	if err != nil {
+		log.Err(err).Msg("Failed to get admin access token")
+		return nil, errors.Internal("Failed to update app key: couldn't get admin access token")
+	}
+	updateAppKeyReq.Header.Add("Authorization", fmt.Sprintf("bearer %s", adminAccessToken))
+	updateAppKeyReq.Header.Add("Content-Type", "application/json")
+
+	updateAppKeyRes, err := ctxhttp.Do(ctx, client, updateAppKeyReq)
+	if err != nil {
+		log.Err(err).Msg("Failed to update app key - failed to make call to gotrue")
+		return nil, errors.Internal("Failed to update app key")
+	}
+	defer updateAppKeyRes.Body.Close()
+
+	if updateAppKeyRes.StatusCode != http.StatusOK {
+		return nil, errors.Internal("Failed to update app key")
```

Comment:
I didn't understand your comment.

Reply:
I mean as you're returning internal error in case the `updateAppKeyRes.StatusCode` then have that code return as well as part of response or log it so that we know what type of status code is returned by gotrue.

Reply:
In case if gotrue returns 401, to user it is still a 500 because it is internal failure - that is why it is always internal error.

Reply:
That's not the point. I am saying how do we know what gotrue returned if it is blindly returning internal failure error here? To debug it either we need to return that code as part of the message or log it. I am not saying don't return internal failure error.

Reply:
ah. fixed.